### PR TITLE
Fix caching issue in poll-cv engine

### LINKED ISSUE #17856: Bazel RBE flake: //test/core/end2end:h2_http_proxy_test@bad_ping@poller=poll-cv
```
Executing tests from //test/core/end2end:h2_http_proxy_test@bad_ping@poller=poll-cv
-----------------------------------------------------------------------------
D0129 09:19:02.646180569      20 test_config.cc:391]         test slowdown factor: sanitizer=1, fixture=1, poller=5, total=5
D0129 09:19:02.646541251      20 ev_posix.cc:170]            Using polling engine: poll-cv
D0129 09:19:02.646614106      20 dns_resolver.cc:334]        Using native dns resolver
I0129 09:19:02.646665281      20 http_proxy_fixture.cc:536]  Proxy address: localhost:29742
D0129 09:19:02.648091481      20 dns_resolver.cc:275]        Start resolving.
I0129 09:19:02.649105964      20 http_connect_handshaker.cc:300] Connecting to server localhost:29741 via HTTP proxy ipv6:[::1]:29742
D0129 09:19:02.649240232      24 http_proxy_fixture.cc:384]  on_read_request_done: 0x7f2168001620 "No Error"
I0129 09:19:02.649886361      20 subchannel.cc:833]          New connected subchannel at 0xffa710 for subchannel 0x7f2164002d70
I0129 09:19:02.679963389      24 http_proxy_fixture.cc:162]  HTTP proxy server read: {"created":"@1548753542.679930674","description":"Socket closed","fd":8,"file":"src/core/lib/iomgr/tcp_posix.cc","file_line":476,"grpc_status":14,"target_address":"ipv6:[::1]:29741"}
I0129 09:19:02.680052327      24 http_proxy_fixture.cc:162]  HTTP proxy client read: {"created":"@1548753542.680021952","description":"FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":532,"grpc_status":14,"referenced_errors":[{"created":"@1548753542.679930674","description":"Socket closed","fd":8,"file":"src/core/lib/iomgr/tcp_posix.cc","file_line":476,"grpc_status":14,"target_address":"ipv6:[::1]:29741"}]}
D0129 09:19:02.680082746      24 http_proxy_fixture.cc:130]  endpoints: 0x7f2168000fd0 0x7f2168005b70
E0129 09:19:02.680162673      20 chttp2_transport.cc:1142]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
D0129 09:19:02.680315921      20 dns_resolver.cc:255]        In cooldown from last resolution (from 32 ms ago). Will resolve again in 968 ms
I0129 09:19:03.680453263      20 http_proxy_fixture.cc:536]  Proxy address: localhost:29744
D0129 09:19:03.681320195      20 dns_resolver.cc:275]        Start resolving.
I0129 09:19:03.681914326      20 http_connect_handshaker.cc:300] Connecting to server localhost:29743 via HTTP proxy ipv6:[::1]:29744
I0129 09:19:23.681684405      28 subchannel.cc:870]          Connect failed: {"created":"@1548753563.681604206","description":"FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":532,"grpc_status":14,"referenced_errors":[{"created":"@1548753563.681535129","description":"Handshake timed out","file":"src/core/lib/channel/handshaker.cc","file_line":290}]}
I0129 09:19:23.682004546      20 http_connect_handshaker.cc:300] Connecting to server localhost:29743 via HTTP proxy ipv4:127.0.0.1:29744
I0129 09:19:43.681707831      28 subchannel.cc:870]          Connect failed: {"created":"@1548753583.681625270","description":"FD shutdown","file":"src/core/lib/iomgr/ev_poll_posix.cc","file_line":532,"grpc_status":14,"referenced_errors":[{"created":"@1548753583.681553123","description":"Handshake timed out","file":"src/core/lib/channel/handshaker.cc","file_line":290}]}
I0129 09:19:43.681790862      28 subchannel.cc:750]          Subchannel 0x7f2164002d70: Retry immediately
I0129 09:19:43.681802491      28 subchannel.cc:711]          Failed to connect to channel, retrying
D0129 09:19:43.681957167      28 dns_resolver.cc:275]        Start resolving.
E0129 09:19:43.682062385      20 cq_verifier.cc:273]         cq returned unexpected event: OP_COMPLETE: tag:0x1 OK
E0129 09:19:43.682079590      20 cq_verifier.cc:280]         expected tags:
tag(101) GRPC_OP_COMPLETE success=1 test/core/end2end/tests/bad_ping.cc:312
*******************************
Caught signal SIGABRT
```

https://source.cloud.google.com/results/invocations/6a93f904-ef02-4412-8b70-affa6063c03e/targets/%2F%2Ftest%2Fcore%2Fend2end:h2_http_proxy_test@bad_ping@poller%3Dpoll-cv/log

### PR DESCRIPTION
**Pull request by yashykt**
poll-cv engine uses background pollers to perform the actual polling. When we close fds, it is possible for active_pollers to still have results for older fds. With this, if we create new fds that get the same file descriptor number as the previously closed fd, then we end up reusing the older results.

To avoid this, when we are done with a fd, we should make sure that we do not end up reusing the older result. This PR introduces a bool flag which is set to false when a cached poller currently polls on a closed fd. When we are looking for cached pollers to reuse, this flag is checked.

Fixes #17856, #17680, #17399 

There might be better ways to do this. I am open to suggestions.

### CODE REVIEW

**Comment by jtattermusch** [comment]
@yashykt, great investigation!

**Comment by jtattermusch** [comment]
The fix seems to be helping - I was able to reproduce before with manual RBE run (`bazel --bazelrc=tools/remote_build/manual.bazelrc test --config=dbg //test/core/end2end:h2_http_proxy_test@bad_ping@poller=poll-cv --runs_per_test=500`) and with the fix I'm not able to reproduce anymore.

**Comment by jtattermusch** [comment]
Looks like this is breaking a lot of tests that use the poll poller - more work might be needed
```
D0306 14:43:19.088866000 140736179336064 test_config.cc:392]           test slowdown factor: sanitizer=1, fixture=1, poller=1, total=1
D0306 14:43:19.090218000 140736179336064 ev_posix.cc:170]              Using polling engine: poll
D0306 14:43:19.090599000 140736179336064 dns_resolver_ares.cc:488]     Using ares dns resolver
D0306 14:43:19.090691000 140736179336064 buffer_list.cc:299]           Timestamps callback is not enabled for this platform
I0306 14:43:19.090722000 140736179336064 endpoint_tests.cc:71]         multiple_shutdown_test/tcp/tcp_socketpair
I0306 14:43:19.091086000 140736179336064 endpoint_tests.cc:71]         read_and_write_test/tcp/tcp_socketpair
D0306 14:43:19.091136000 140736179336064 endpoint_tests.cc:180]        num_bytes=10000000 write_size=100000 slice_size=8192 shutdown=0
I0306 14:43:19.091155000 140736179336064 endpoint_tests.cc:188]        Start read and write test with 10000000 bytes, slice size 8192
E0306 14:43:19.091829000 140736179336064 sync_posix.cc:103]            assertion failed: pthread_mutex_lock(mu) == 0
```

**Comment by yashykt** [comment]
I don't get why pthread_mutex_lock is failing for macos

File: src/core/lib/iomgr/ev_poll_posix.cc [hunk 1]
```diff
@@ -395,6 +399,28 @@ static void unref_by(grpc_fd* fd, int n) {
 #endif
   gpr_atm old = gpr_atm_full_fetch_add(&fd->refst, -n);
   if (old == n) {
+    /* We have a cache of pollers and their results. Even though we are done
+     * with this fd, we might get an fd with the same number. To avoid using
+     * results from this fd for a future fd, zero out the results for this fd.
+     */
+    gpr_mu_lock(&g_cvfds.mu);
```
Comment (kpayson64) [inline src/core/lib/iomgr/ev_poll_posix.cc]: We should only do this if we are using the poll-cv engine.  The g_cvfds mutex is only initialized if we are using the poll-cv engine I believe.

Comment (yashykt) [reply src/core/lib/iomgr/ev_poll_posix.cc]: I just realized that :)

File: src/core/lib/iomgr/ev_poll_posix.cc [hunk 2]
```diff
@@ -358,9 +362,12 @@ static void fork_fd_list_add_wakeup_fd(grpc_cached_wakeup_fd* fd) {
   }
 }
 
-  /*******************************************************************************
-   * fd_posix.c
-   */
+/*******************************************************************************
+ * fd_posix.c
+ */
```
Comment (vjpai) [inline src/core/lib/iomgr/ev_poll_posix.cc]: Why is that comment present at all?

**Comment by jtattermusch** [comment]
poll-cv was removed in #18298. Close?

**Pull request closed without merging**